File: cpp/src/2021/apr/partition-list.cc
#include "common.h"

class Solution
{
public:
    ListNode *partition(ListNode *head, int x)
    {
        ListNode node1(0), node2(0);
        ListNode *p1 = &node1, *p2 = &node2;

        while (head)
        {
            if (head->val < x)
            {
                p1 = p1->next = head;
            }
            else
            {
                p2 = p2->next = head;
            }
            head = head->next;
        }
        p2->next = nullptr;
        p1->next = node2.next;
        return node1.next;
    }
};

File: code/java/src/main/java/cn/az/code/year2021/apr/PartitionList.java
package cn.az.code.year2021.apr;

import cn.az.code.common.ListNode;

/**
 * @author ycpang
 * @since 2021-04-14 16:24
 */
public class PartitionList {

    public ListNode partition(ListNode head, int x) {
        ListNode node1 = new ListNode(0), node2 = new ListNode(0);
        ListNode n1 = node1, n2 = node2;

        while (head != null) {
            if (head.val < x) {
                n1 = n1.next = head;
            } else {
                n2 = n2.next = head;
            }
            head = head.next;
        }
        n1.next = node2.next;
        n2.next = null;
        return node1.next;
    }
}
